File: fe/src/api/category.api.ts
import apiClient from '@/lib/api-client'
import adminApiClient from '@/lib/admin-api-client'
import type { ApiResponse, CategoryRequest, CategoryResponse } from '@/types/api'

const CATEGORY_BASE_URL = '/api/categories'
const ADMIN_CATEGORY_URL = '/api/admin/categories'

export const categoryApi = {
  getAll: async (): Promise<CategoryResponse[]> => {
    const response = await apiClient.get<ApiResponse<CategoryResponse[]>>(CATEGORY_BASE_URL)
    return response.data.data
  },

  getById: async (id: number): Promise<CategoryResponse> => {
    const response = await apiClient.get<ApiResponse<CategoryResponse>>(`${CATEGORY_BASE_URL}/${id}`)
    return response.data.data
  },

  getBySlug: async (slug: string): Promise<CategoryResponse> => {
    const response = await apiClient.get<ApiResponse<CategoryResponse>>(`${CATEGORY_BASE_URL}/slug/${slug}`)
    return response.data.data
  },

  getByParentId: async (parentId: number): Promise<CategoryResponse[]> => {
    const response = await apiClient.get<ApiResponse<CategoryResponse[]>>(`${CATEGORY_BASE_URL}/parent/${parentId}`)
    return response.data.data
  },

  // Admin: Create category (calls Server port 8081)
  create: async (data: CategoryRequest): Promise<CategoryResponse> => {
    const response = await adminApiClient.post<ApiResponse<CategoryResponse>>(ADMIN_CATEGORY_URL, data)
    return response.data.data
  },

  // Admin: Update category (calls Server port 8081)
  update: async (id: number, data: CategoryRequest): Promise<CategoryResponse> => {
    const response = await adminApiClient.put<ApiResponse<CategoryResponse>>(`${ADMIN_CATEGORY_URL}/${id}`, data)
    return response.data.data
  },

  // Admin: Delete category (calls Server port 8081)
  delete: async (id: number): Promise<void> => {
    await adminApiClient.delete<ApiResponse<void>>(`${ADMIN_CATEGORY_URL}/${id}`)
  },
}

File: be/landing/src/main/java/com/example/spring_ecom/controller/api/category/CategoryAPI.java
package com.example.spring_ecom.controller.api.category;

import com.example.spring_ecom.controller.api.category.model.CategoryRequest;
import com.example.spring_ecom.controller.api.category.model.CategoryResponse;
import com.example.spring_ecom.core.response.ApiResponse;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.validation.Valid;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@Tag(name = "Category", description = "Category management APIs")
@RequestMapping("/api/categories")
public interface CategoryAPI {
    
    @Operation(summary = "Get all categories")
    @GetMapping
    ApiResponse<List<CategoryResponse>> getAllCategories();
    
    @Operation(summary = "Get category by id")
    @GetMapping("/{id}")
    ApiResponse<CategoryResponse> getCategoryById(@PathVariable Long id);
    
    @Operation(summary = "Get category by slug")
    @GetMapping("/slug/{slug}")
    ApiResponse<CategoryResponse> getCategoryBySlug(@PathVariable String slug);
    
    @Operation(summary = "Get categories by parent id")
    @GetMapping("/parent/{parentId}")
    ApiResponse<List<CategoryResponse>> getCategoriesByParentId(@PathVariable Long parentId);
    
    @Operation(summary = "Create new category")
    @PostMapping
    ApiResponse<CategoryResponse> createCategory(@Valid @RequestBody CategoryRequest request);
    
    @Operation(summary = "Update category")
    @PutMapping("/{id}")
    ApiResponse<CategoryResponse> updateCategory(@PathVariable Long id, @Valid @RequestBody CategoryRequest request);
    
    @Operation(summary = "Delete category")
    @DeleteMapping("/{id}")
    ApiResponse<Void> deleteCategory(@PathVariable Long id);
}
